File: src/shared/week.ts
import { localDate } from './format';

/**
 * Week math for gym streaks. Weeks are local and start on Monday; a week is
 * identified by its Monday's local date ('YYYY-MM-DD').
 */

/** Monday of the week containing `date` */
export function weekKey(date = new Date()): string {
  const d = new Date(date);
  d.setDate(d.getDate() - ((d.getDay() + 6) % 7)); // Mon=0 … Sun=6
  return localDate(d);
}

export function prevWeekKey(key: string): string {
  const [y, m, d] = key.split('-').map(Number);
  const date = new Date(y, m - 1, d);
  date.setDate(date.getDate() - 7);
  return localDate(date);
}

/** The 7 local dates (Mon–Sun) of the week identified by `key` */
export function weekDates(key: string): string[] {
  const [y, m, d] = key.split('-').map(Number);
  const monday = new Date(y, m - 1, d);
  return Array.from({ length: 7 }, (_, i) => {
    const day = new Date(monday);
    day.setDate(monday.getDate() + i);
    return localDate(day);
  });
}

export function countInWeek(checkins: Record<string, number>, key: string): number {
  return weekDates(key).filter((date) => date in checkins).length;
}

/**
 * Whether local `now` falls inside the [start, end) quiet window. A start
 * after the end wraps overnight (22:00→08:00). Equal bounds = never quiet.
 */
export function inQuietHours(startHhmm: string, endHhmm: string, now = new Date()): boolean {
  const toMin = (hhmm: string): number | null => {
    const match = /^(\d{1,2}):(\d{2})$/.exec(hhmm.trim());
    if (!match) return null;
    const h = Number(match[1]);
    const m = Number(match[2]);
    return h > 23 || m > 59 ? null : h * 60 + m;
  };
  const start = toMin(startHhmm);
  const end = toMin(endHhmm);
  if (start === null || end === null || start === end) return false;
  const cur = now.getHours() * 60 + now.getMinutes();
  return start < end ? cur >= start && cur < end : cur >= start || cur < end;
}

/** Next occurrence of a local 'HH:MM' wall-clock time, as ms epoch */
export function nextDailyOccurrence(hhmm: string, now = new Date()): number {
  const [h, min] = hhmm.split(':').map(Number);
  const next = new Date(now);
  next.setHours(h, min, 0, 0);
  if (next.getTime() <= now.getTime()) {
    next.setDate(next.getDate() + 1);
  }
  return next.getTime();
}

File: ios/ADHDReaderCore/Sources/ADHDReaderCore/Week.swift
import Foundation

/// Week math for gym streaks, ported from `src/shared/week.ts`. Weeks are local
/// and start on Monday; a week is identified by its Monday's local 'YYYY-MM-DD'.
public enum Week {
    /// Monday of the week containing `date`.
    public static func weekKey(_ date: Date = Date()) -> String {
        let cal = Calendar.current
        let jsDay = cal.component(.weekday, from: date) - 1 // Sun=0 … Sat=6
        let offset = (jsDay + 6) % 7                        // Mon=0 … Sun=6
        let monday = cal.date(byAdding: .day, value: -offset, to: date)!
        return DateUtil.localDate(monday)
    }

    public static func prevWeekKey(_ key: String) -> String {
        let date = DateUtil.parseLocalDate(key)!
        return DateUtil.localDate(Calendar.current.date(byAdding: .day, value: -7, to: date)!)
    }

    /// The 7 local dates (Mon–Sun) of the week identified by `key`.
    public static func weekDates(_ key: String) -> [String] {
        let monday = DateUtil.parseLocalDate(key)!
        return (0..<7).map {
            DateUtil.localDate(Calendar.current.date(byAdding: .day, value: $0, to: monday)!)
        }
    }

    public static func countInWeek(_ checkins: [String: Int], _ key: String) -> Int {
        weekDates(key).filter { checkins[$0] != nil }.count
    }

    /// Next occurrence of a local 'HH:MM' wall-clock time, as ms epoch.
    public static func nextDailyOccurrence(_ hhmm: String, now: Date = Date()) -> Millis {
        let parts = hhmm.split(separator: ":").compactMap { Int($0) }
        let cal = Calendar.current
        var next = cal.date(bySettingHour: parts[0], minute: parts[1], second: 0, of: now)!
        if next <= now { next = cal.date(byAdding: .day, value: 1, to: next)! }
        return Int((next.timeIntervalSince1970 * 1000).rounded())
    }
}
